### PR TITLE
Feature/audit logging

### PR DESCRIPTION
**Pull request by fieldju**
- Creates an optional audit log that is disabled by default that will log messages in a flattened JSON object that is Athena friendly.
- If the properties are set in environment.properties the logback configuration will automatically copy the audit logs that get rolled every 5 mintues to S3

example if the following are set

```
cms.event.processors.com.nike.cerberus.event.processor.AuditLogProcessor=true
cms.audit.bucket="athena-test-xxxxxx"
cms.audit.bucket_region="us-west-2"
```

Then the processor and s3 copying will be enabled.

un-minified example of audit log message
```json
{
  "event-timestamp":"2018-01-26 11:35:35",
  "principal-name":"arn:aws:iam::xxxxxxx:role/dev-cerberus-health-check-CerberusHealthCheckRole-xxxxxxx",
  "principal-type":"IAM",
  "principal-token-created":"2018-01-26 11:35:35",
  "principal-token-expires":"2018-01-27 12:35:35",
  "principal-is-admin":"false",
  "ip-address":"22.22.22.22",
  "x-forwarded-for":"33.33.33.33",
  "client-version":"Unknown",
  "http-method":"GET",
  "path":"/v1/secret/app/health-check-bucket/healthcheck",
  "action":"Reading secret with path: /v1/secret/app/health-check-bucket/healthcheck",
  "name":"ReadSecureData Endpoint Called",
  "sdb-name-slug":"health-check-bucket",
  "originating-class":"ReadSecureData"
}
```

The intent is to create a command in the CLI that will create the bucket and configure the Athena table and enable the processor.

### CODE REVIEW

File: src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        if (!rolledLogFile.exists() || rolledLogFile.length() == 0) {
+            return;
+        }
+
+        executor.execute(() -> {
+            try {
+                getAmazonS3().putObject(getBucket(), ENV_NAME + "/audit-logs/" + rolledLogFile.getName(), rolledLogFile);
+                rolledLogFile.delete();
+            } catch (Exception ex) {
+                ex.printStackTrace();
```
Comment (tlisonbee) [inline src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java]: shouldn't this be a logger?

Comment (tlisonbee) [reply src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java]: Also, do we want to catch here?

File: src/main/java/com/nike/cerberus/endpoints/CustomizableAuditData.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nike.cerberus.endpoints;
+
+public class CustomizableAuditData {
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/endpoints/CustomizableAuditData.java]: javadoc

File: src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java (hunk 1 above)
Comment (fieldju) [reply src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java]: You don't have access to the logger in side the log back internals

File: src/main/java/com/nike/cerberus/event/processor/AuditLogProcessor.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                    .put("action", event.getAction())
+                    .put("name", event.getName())
+                    .put("sdb-name-slug", event.getSdbNameSlug())
+                    .put("originating-class", event.getOriginatingClass())
+                    .build();
+
+            try {
+                auditLogger.info(om.writeValueAsString(flattenedAuditEvent));
+            } catch (JsonProcessingException e) {
+                LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME).error("failed to log audit event", e);
```
Comment (tlisonbee) [inline src/main/java/com/nike/cerberus/event/processor/AuditLogProcessor.java]: Is that really the logger you want?

Comment (fieldju) [reply src/main/java/com/nike/cerberus/event/processor/AuditLogProcessor.java]: Yes, want the errors going to the main log appender and not the audit log

File: src/main/resources/logback.groovy [hunk 4]
```diff
@@ -16,15 +16,14 @@
...
-import ch.qos.logback.core.rolling.FixedWindowRollingPolicy
-import ch.qos.logback.core.rolling.RollingFileAppender
-import ch.qos.logback.core.rolling.SizeBasedTriggeringPolicy
+import ch.qos.logback.core.util.FileSize
+import com.nike.cerberus.util.ConfigUtils
 
 import static ch.qos.logback.classic.Level.*
 import static ch.qos.logback.core.spi.ContextAware.addInfo
 
+System.setProperty("logback.debug", "true")
```
Comment (tlisonbee) [inline src/main/resources/logback.groovy]: We don't want debug do we?

File: src/main/java/com/nike/cerberus/event/processor/AuditLogProcessor.java (hunk 3 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/event/processor/AuditLogProcessor.java]: I mean "ROOT" is fine with me, but It can be any string other than this.getClass()

File: src/main/resources/logback.groovy (hunk 4 above)
Comment (fieldju) [reply src/main/resources/logback.groovy]: I just enabled this so I can get more info in dev2 and bake an AMI

File: src/main/resources/logback.groovy [hunk 5]
```diff
@@ -193,8 +192,7 @@ logger("com.nike.metrics.newrelic.NewRelicReporter", WARN, allAsyncAppendersArra
 
 logger("com.nike.cerberus", INFO, allAsyncAppendersArray, false)
 
-// TODO revert before merge to master
-logger("com.nike.cerberus.jobs.HystrixMetricsProcessingJob", DEBUG, allAsyncAppendersArray, false)
+logger("com.nike.cerberus.jobs.HystrixMetricsProcessingJob", INFO, allAsyncAppendersArray, false)
```
Comment (tlisonbee) [inline src/main/resources/logback.groovy]: What was "revert before merge to master"?

Are there other bad changes in here?

Is this highlander?  Things we need to update?

**Comment by tlisonbee** [comment]
This seems worth performance testing before merging.

I definitely would not want to merge till after next Monday.

File: src/main/resources/logback.groovy (hunk 5 above)
Comment (fieldju) [reply src/main/resources/logback.groovy]: That was referencing the LOG level of DEBUG, I switched it back to INFO

File: src/main/resources/logback.groovy (hunk 4 above)
Comment (tlisonbee) [reply src/main/resources/logback.groovy]: Are you removing then?

File: src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java (hunk 1 above)
Comment (tlisonbee) [reply src/main/java/ch/qos/logback/core/rolling/AuditLogsS3FixedWindowRollingPolicy.java]: If that is the case, how about also adding a System.err with error message?  Also, maybe a comment?

**Comment by fieldju** [comment]
Yeah, I agree about performance testing it, it is all async so it shouldn't have an impact, but I will verify. 

I am also verifying that the log rotation to S3 is working as I expect.

**Comment by tlisonbee** [comment]
Also, maybe something needs to be added to docs or README?

**Comment by tlisonbee** [comment]
It seems like the kind of change that doesn't hurt to let sit in a test environment for awhile in case issues show up: memory leaks, edge cases, etc.

I went through the code quickly but I also wonder about error scenarios, I'm guessing S3 client has some retry built-in, but also what happens if S3 is down for 2 hours?

**Comment by fieldju** [comment]
Yeah letting it sit for a bit sounds reasonable. I am still UAT'ing the log rolling pieces to make sure everything is working as I expect, once I get that taken care of I will remove that debug statement and do a perf test.

As for S3 being down, as it is right now and I am open to idea's the AuditLogsS3FixedWindowRollingPolicy will attempt to upload the rolled file and then delete the local copy, if some part of that fails it just moves on and never attempts to do anything else with that file.
so if there is a network blip you lose 5 minutes of logs.

logback is configured to keep 100 rolled logs around before it deletes things itself, so in an S3 outage there would be 100 * 5 = 8.3 hours or 10gb (which ever is first) worth of logs around before they start getting deleted locally, with manual intervention required to get them in the S3 bucket when it comes back online.

Lets talk more Monday in person.

Cheers

File: src/main/resources/logback.groovy (hunk 4 above)
Comment (fieldju) [reply src/main/resources/logback.groovy]: After perf testing, you have to have this enabled if you want any sort of logging or std out/err from these classes.
